src/integrationTest/java/org/opensearch/test/framework/audit/AuditLogsRule.java

```diff
@@ -0,0 +1,131 @@
+/*
+* Copyright OpenSearch Contributors
+* SPDX-License-Identifier: Apache-2.0
+*
+* The OpenSearch Contributors require contributions made to
+* this file be licensed under the Apache-2.0 license or a
+* compatible open source license.
+*
+*/
+package org.opensearch.test.framework.audit;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.awaitility.Awaitility;
+import org.awaitility.core.ConditionTimeoutException;
+import org.hamcrest.Matcher;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import org.opensearch.security.auditlog.impl.AuditMessage;
+
+import static java.util.Collections.synchronizedList;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.opensearch.test.framework.matcher.AuditMessageMatchers.atLeastCertainNumberOfAuditsFulfillPredicate;
+import static org.opensearch.test.framework.matcher.AuditMessageMatchers.exactNumberOfAuditsFulfillPredicate;
+
+public class AuditLogsRule implements TestRule {
+
+	private static final Logger log = LogManager.getLogger(AuditLogsRule.class);
+
+	private List<AuditMessage> currentTestAuditMessages;
+
+	public void waitForAuditLogs() {
+		try {
+			TimeUnit.SECONDS.sleep(3);
+			afterWaitingForAuditLogs();
+		} catch (InterruptedException e) {
+			throw new RuntimeException("Waiting for audit logs interrupted.", e);
+		}
+	}
+
+	private void afterWaitingForAuditLogs() {
+		if(log.isDebugEnabled()) {
+			log.debug("Audit records captured during test:\n{}", auditMessagesToString(currentTestAuditMessages));
+		}
+	}
+
+	public void assertExactlyOne(Predicate<AuditMessage> predicate) {
+		assertExactly(1, predicate);
+	}
+
+	public void assertAuditLogsCount(int from, int to) {
+		int actualCount = currentTestAuditMessages.size();
+		String message = "Expected audit log count is between " + from + " and " + to + " but was " + actualCount;
+		assertThat(message, actualCount, allOf(greaterThanOrEqualTo(from), lessThanOrEqualTo(to)));
+	}
+
+	public void assertExactly(long expectedNumberOfAuditMessages, Predicate<AuditMessage> predicate) {
+		assertExactly(exactNumberOfAuditsFulfillPredicate(expectedNumberOfAuditMessages, predicate));
+	}
+
+	private void assertExactly(Matcher<List<AuditMessage>> matcher) {
+		//pollDelay - initial delay before first evaluation
+		Awaitility.await("Await for audit logs")
+			.atMost(3, TimeUnit.SECONDS).pollDelay(0, TimeUnit.MICROSECONDS).until(() -> {
+				return new ArrayList<>(currentTestAuditMessages);
+
+		}, matcher);
```

Comment:
It looks like this pooler maxing the CPU while the cluster is trying to write the logs, if so I think this might impact system under test stability.  It looks like a signal/wait pattern would be more appropriate, as we implemented in the other test cases TestAuditlogImpl.java
https://github.com/opensearch-project/security/blob/16b16765b2940ffb5ac4dcbfc35db959ff4411ee/src/test/java/org/opensearch/security/auditlog/integration/TestAuditlogImpl.java#L60-L83

Reply:
Actually, it is quite advanced active waiting. The first check is performed immediate `pollDelay(0, TimeUnit.MICROSECONDS)`. The default interval between subsequent condition (`pollInterval`) evaluations is equal to 100 ms. Moreover, each condition evaluation is related to searching elements in quite a small list. The list usually contains a couple of elements, up to ~30.

Please let me know what you think.

Reply:
I would check the CI results but as main is busted so I don't have any clean results to inspect.  Would you be able to do a before/after runtime analysis locally and post it here?

Since no new tests were added we should just see the overhead induced by these checks, with these numbers it would be easy to conclude if we would like to dig in vs keeping them exactly as you've build them.

Reply:
I checked this and it seems that there is no measurable difference between tests with and without audit logs in terms of test execution time.